standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java

```diff
@@ -2836,14 +2836,32 @@ long getPartsFound() {
   @Override
   public SQLAllTableConstraints getAllTableConstraints(String catName, String dbName, String tblName)
       throws MetaException, NoSuchObjectException {
-    SQLAllTableConstraints sqlAllTableConstraints = new SQLAllTableConstraints();
-    sqlAllTableConstraints.setPrimaryKeys(getPrimaryKeys(catName, dbName, tblName));
-    sqlAllTableConstraints.setForeignKeys(getForeignKeys(catName, null, null, dbName, tblName));
-    sqlAllTableConstraints.setUniqueConstraints(getUniqueConstraints(catName, dbName, tblName));
-    sqlAllTableConstraints.setDefaultConstraints(getDefaultConstraints(catName, dbName, tblName));
-    sqlAllTableConstraints.setCheckConstraints(getCheckConstraints(catName, dbName, tblName));
-    sqlAllTableConstraints.setNotNullConstraints(getNotNullConstraints(catName, dbName, tblName));
-    return sqlAllTableConstraints;
+
+    catName = StringUtils.normalizeIdentifier(catName);
+    dbName = StringUtils.normalizeIdentifier(dbName);
+    tblName = StringUtils.normalizeIdentifier(tblName);
+    if (!shouldCacheTable(catName, dbName, tblName) || (canUseEvents && rawStore.isActiveTransaction())) {
+      return rawStore.getAllTableConstraints(catName, dbName, tblName);
+    }
+
+    Table tbl = sharedCache.getTableFromCache(catName, dbName, tblName);
+    if (tbl == null) {
+      // The table containing the constraints is not yet loaded in cache
+      return rawStore.getAllTableConstraints(catName, dbName, tblName);
+    }
+    SQLAllTableConstraints constraints = sharedCache.listCachedAllTableConstraints(catName, dbName, tblName);
+
+    // if any of the constraint value is missing then there might be the case of partial constraints are stored in cached.
+    // So fall back to raw store for correct values
+    if (constraints != null && CollectionUtils.isNotEmpty(constraints.getPrimaryKeys()) && CollectionUtils
```

Comment:
Majority of the calls are likely to be redirected to RawStore and lose the advantage of cache. How about an optional flag in SQLAllTableConstraints and TableWrapper to mark if it is a complete snapshot of constraints?

Reply:
+1.

Reply:
Also, what if the table just a primary keys and not other constraints?

nonEmpty(otherConstraints) will return true and we will not cache anything?

Reply:
Good idea. I will check how we can implement a flag.

Reply:
Adding a flag required bigger change. So for now I am reducing the scope of this PR to optimise following
1. Check only once if table exit in cached store.
2. Instead of calling individual constraint in cached store. Add a method which return all constraint at once and if data is not consistent then fall back to rawstore.

Reply:
Why is it hard to set a flag for consistency? If we hit else flow in this case, we shall try to update the cache and if that fails for some reason, then set the flag to false else set to true. Same with pre-warm and refresh flows.
Also, we need to check only the flag to validate the consistency of cache. No need to check for null or empty constraints.

Reply:
Added flag